src/reducers/bookmark.ts

```diff
@@ -0,0 +1,52 @@
+import {Action, BOOKMARK_ADD_PLOT, BOOKMARK_MODIFY_NOTE, BOOKMARK_REMOVE_PLOT} from '../actions';
+import {Bookmark, BookmarkItem} from '../models';
+
+
+export function bookmarkReducer(bookmark: Bookmark, action: Action): Bookmark {
+  const newBookmark = {
+    dict: {
+      ...bookmark.dict
+    },
+    numBookmarks: bookmark.numBookmarks
+  };
+
+  switch (action.type) {
+    case BOOKMARK_ADD_PLOT: {
+      const {plot} = action.payload;
+
+      const specKey = JSON.stringify(plot.spec);
+      if (!bookmark.dict[specKey]) {
+        const bookmarkItem: BookmarkItem = {
+          plot: plot,
+          note: '',
+        };
+
+        newBookmark.numBookmarks++;
```

Comment:
This stills read like causing side-effect (but to a new object).  

Why don't you just follow the pattern that we generally use throughout the project.

Basically, get rid of your `const newBookmark` 

and just return a new object based on spread.

For example, here you can just do

```
return {
   dict: {
     ...dict,
     [specKey]: bookmarkItem
   },
   numBookmarks: numBookmarks + 1
} 
```

(Add `const {numBookmarks, dict} = bookmark;` to the top.)